File: Famous Algorithms/euclidean_algorithm.py
#extended version of Euclid's algorithm to find GCD of 2 numbers
#runs in O(log N) time/space complexity for GCD of numbers a and b 
#in comparison, the standard Euclidean algorithm runs in O (log (min (a,b)))

def euclidExtended(a,b):
    if a==0:
        return b,0, 1 
    result, a1, b1 = euclidExtended(b%a,a)
    
    a2 = b1- (b//a) *a1
    b2 = a1
    return result, a2, b2 #used as input to recursive call, 

#example driver, change a and b as desired

a,b = 45,10
g,x,y = euclidExtended(a,b)
print("gcd of", a,"," ,b ,"is equal to: " , g)
    
    
        

File: Famous Algorithms/euclidean_algorithm.java
//extended version of Euclid's algorithm to find GCD of 2 numbers
//runs in O(log N) time/space complexity for GCD of numbers a and b 
//in comparison, the standard Euclidean algorithm runs in O (log (min (a,b)))

public class euclidean_algorithm {
    public static int euclid(int a, int b, int c, int d){
        if(a == 0){
            c = 0;
            d = 0;
            return b;
        }
        int c1 = 1,d1 = 1;
        int result = euclid(b%a, a, c1, d1);
        //update with recursive call
        c = d1 - (b / a) * c1;
        d = c1;
        return result;
    }
    //driver
    public static void main(String[] args) {
        int c =1, d = 1;
        int a = 45;
        int b = 10;
        int gcd = euclid(a, b, c, d);
        System.out.print("gcd of "+ a+"," +b +" is equal to: " + gcd);
    }
}
